packages/cdktf-cli/lib/cdktf-stack.ts

```diff
@@ -266,83 +132,201 @@ export class CdktfStack {
   public get isRunning(): boolean {
     return !this.isPending && !this.isDone;
   }
-  public get currentWorkPromise(): Promise<void> | undefined {
-    if (!this.isRunning) {
-      return undefined;
-    }
 
-    return this.waitOnMachineDone();
+  private notifyState(
+    update:
+      | StackUpdate
+      | StackApprovalUpdate
+      | { type: "idle" }
+      | { type: "done" }
+      | { type: "error" }
+  ) {
+    logger.debug(`[${this.stackName}]: ${update.type}`);
+    this.currentState = update.type;
+    switch (update.type) {
+      case "idle":
+      case "done":
+      case "error":
+        break;
+
+      case "outputs fetched":
+      case "deployed":
+        logger.debug(`Outputs: ${JSON.stringify(update.outputs)}`);
+        logger.debug(
+          `OutputsByConstructId: ${JSON.stringify(update.outputsByConstructId)}`
+        );
+        this.outputs = update.outputs;
+        this.outputsByConstructId = update.outputsByConstructId;
+        this.context.onUpdate(update);
+        break;
+
+      default:
+        this.context.onUpdate(update);
+        break;
+    }
   }
 
-  private waitOnMachineDone(): Promise<void> {
-    return new Promise((resolve, reject) => {
-      if (this.currentState === "done") {
-        resolve();
-      }
-
-      if (this.currentState === "error") {
-        reject(new Error(this.stateMachine.state.context.message));
+  private logCallback(
+    stateName: string
+  ): (message: string, isError?: boolean) => void {
```

Comment:
In `TerraformCloud` this is defined as:
```
private readonly createLogSender = (_phase: string) => (_stdout: string, _isErr = false) => {}
```
In `TerraformCli` this is defined as:
```
onLog = (_stateName: string) => (_stdout: string, _isErr = false) => {}
```
I personally think that `phase` is the right term to not confuse this with the state / current state of this class.
I'd opt to name all methods / arguments `createLogSender` due to this being a higher order function.
For the other args I vote for `message` and `isError`.

Reply:
After reading more of the code, I noticed that this does not handle just any logs but logs emitted by Terraform (either the CLI or the CLI running in TFC). So, I'd propose to make this clear.
Something along the lines of: `createTerraformLogHandler` even, because that is what it does in the end (afaik). Or am I missing something?